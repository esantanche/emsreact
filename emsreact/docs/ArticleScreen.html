<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ArticleScreen</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ArticleScreen</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ArticleScreen<span class="signature">()</span><span class="type-signature"></span></h2>
        
            <div class="class-description">Screen showing a single article. The text can contain html tags. In this case we render them.</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="ArticleScreen"><span class="type-signature"></span>new ArticleScreen<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="article_ArticleScreen.js.html">article/ArticleScreen.js</a>, <a href="article_ArticleScreen.js.html#line22">line 22</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="componentDidMount"><span class="type-signature"></span>componentDidMount<span class="signature">()</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    When the component mounts, we fetch the article we have to display.

The url that triggers displaying this Screen is like this:
/article/:articlenid

For example it can be: /article/27

The number is the node id of the article in the Drupal backend.

This component uses the service ArticleService for it to fetch
that specific article.
</div>













<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="article_ArticleScreen.js.html">article/ArticleScreen.js</a>, <a href="article_ArticleScreen.js.html#line44">line 44</a>
    </li></ul></dd>
    

    

    

    
</dl>


















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a></li><li><a href="ArticleCard.html">ArticleCard</a></li><li><a href="ArticleScreen.html">ArticleScreen</a></li><li><a href="ColouredCard.html">ColouredCard</a></li><li><a href="MasterAppBar.html">MasterAppBar</a></li><li><a href="SocialCard.html">SocialCard</a></li></ul><h3>Global</h3><ul><li><a href="global.html#theme">theme</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Thu Aug 02 2018 12:45:20 GMT+0000 (UTC)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>