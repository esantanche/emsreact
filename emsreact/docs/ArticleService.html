<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: ArticleService</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: ArticleService</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>ArticleService<span class="signature">()</span><span class="type-signature"></span></h2>
        
            <div class="class-description"><p>This service fetches articles from the backend.</p></div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="ArticleService"><span class="type-signature"></span>new ArticleService<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="services_article_ArticleService.js.html">services/article/ArticleService.js</a>, <a href="services_article_ArticleService.js.html#line10">line 10</a>
    </li></ul></dd>
    

    

    

    
</dl>


















    
    </div>

    

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    
    <h4 class="name" id="fetch_articles"><span class="type-signature"></span>fetch_articles<span class="signature">(filters, callback_to_return_articles)</span><span class="type-signature"></span></h4>
    

    



<div class="description">
    <p>Function fetching articles.</p>
<p>There are three implemented filters.</p>
<p>The first one fetches a single article. It needs the node id the backend assigned to the article.</p>
<p>The second fetches all articles for a given topic. Topics can be &quot;How I work&quot;, &quot;The tools I use&quot;, etc.</p>
<p>The third fetches sticky articles only or all af them.</p>
<p>It's possible to combine the topic and sticky filters. Of course the nid filter (single article) has to be used alone.</p>
</div>









    <h5>Parameters:</h5>
    

<table class="params">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>filters</code></td>
            

            <td class="type">
            
                
<span class="param-type">object</span>


            
            </td>

            

            

            <td class="description last"><p>Filters to use when fetching articles</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>callback_to_return_articles</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last"><p>Function to call to return the fetched articles</p></td>
        </tr>

    
    </tbody>
</table>






<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="services_article_ArticleService.js.html">services/article/ArticleService.js</a>, <a href="services_article_ArticleService.js.html#line28">line 28</a>
    </li></ul></dd>
    

    

    

    
</dl>


















        
    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="AlignedGridPane.html">AlignedGridPane</a></li><li><a href="App.html">App</a></li><li><a href="ArticleCard.html">ArticleCard</a></li><li><a href="ArticleScreen.html">ArticleScreen</a></li><li><a href="ArticleService.html">ArticleService</a></li><li><a href="BodyPane.html">BodyPane</a></li><li><a href="BottomGridPane.html">BottomGridPane</a></li><li><a href="ColouredCard.html">ColouredCard</a></li><li><a href="FooterScreen.html">FooterScreen</a></li><li><a href="FrontGridPane.html">FrontGridPane</a></li><li><a href="HeaderMenu.html">HeaderMenu</a></li><li><a href="HeaderScreen.html">HeaderScreen</a></li><li><a href="HomeButtonPane.html">HomeButtonPane</a></li><li><a href="HomeButtonText.html">HomeButtonText</a></li><li><a href="HomeScreen.html">HomeScreen</a></li><li><a href="InnerGridPane.html">InnerGridPane</a></li><li><a href="ListOfArticlesScreen.html">ListOfArticlesScreen</a></li><li><a href="MasterAppBar.html">MasterAppBar</a></li><li><a href="MasterAvatar.html">MasterAvatar</a></li><li><a href="MenuButton.html">MenuButton</a></li><li><a href="MessageDialog.html">MessageDialog</a></li><li><a href="PickMyBrainDialog.html">PickMyBrainDialog</a></li><li><a href="PickMyBrainMessageService.html">PickMyBrainMessageService</a></li><li><a href="PickMyBrainScreen.html">PickMyBrainScreen</a></li><li><a href="SeparatorPane.html">SeparatorPane</a></li><li><a href="SocialCard.html">SocialCard</a></li><li><a href="SocialScreen.html">SocialScreen</a></li><li><a href="TopicScreen.html">TopicScreen</a></li><li><a href="TopicService.html">TopicService</a></li></ul><h3>Global</h3><ul><li><a href="global.html#theme">theme</a></li><li><a href="global.html#title_to_slug">title_to_slug</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue Aug 21 2018 07:30:21 GMT+0000 (UTC)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>