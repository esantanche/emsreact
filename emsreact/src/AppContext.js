// This is the application context
// We populate it in App.js and provide it to the entire application

import React from "react";

const AppContext = React.createContext(null);

export default AppContext;