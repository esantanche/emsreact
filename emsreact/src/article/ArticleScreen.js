import React, { Component } from 'react';

// Material UI components
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';

// Material UI icons
import Clear from '@material-ui/icons/Clear';
import IconButton from '@material-ui/core/IconButton';

// From components library
import FrontGridPane from '../components_library/panes/FrontGridPane';

// The injector allows this component to use services, see README.md in folder 'services'
import { injector } from 'react-services-injector';

/**
 * Screen showing a single article. The text can contain html tags. In this case we render them.
 */
class ArticleScreen extends Component {

    constructor(props) {
        super(props);

        this.state = { article: [ { } ] };

    }

    /**
     * When the component mounts, we fetch the article we have to display.
     *
     * The url that triggers displaying this Screen is like this:
     * /article/:articlenid
     *
     * For example it can be: /article/27
     *
     * The number is the node id of the article in the Drupal backend.
     *
     * This component uses the service ArticleService for it to fetch
     * that specific article.
     */
    componentDidMount() {

        if (!this.props.match.params.articlenid) {

            console.error("ArticleScreen::componentDidMount");
            console.error("this.props.match.params.articlenid is not defined");
            console.error("This is a bug");

            return;
        }

        const { ArticleService } = this.services;

        this.ArticleService = ArticleService;

        var self = this;

        this.ArticleService.fetch_articles({ nid: this.props.match.params.articlenid },
            function (article) {

                self.setState({ article: article });

            }

        )

    }

    render() {

        return (

            <React.Fragment>

                <FrontGridPane>

                    <Card>

                        <CardHeader title={this.state.article[0].title}
                                    action={
                                        <IconButton href="/">
                                            <Clear />
                                        </IconButton>
                                    }/>

                        <CardContent>

                            <Typography variant="body1" component="div">

                                <div dangerouslySetInnerHTML={{ __html: this.state.article[0].body }} />

                            </Typography>

                        </CardContent>

                    </Card>

                </FrontGridPane>

            </React.Fragment>

        )

    }

}

export default injector.connect(ArticleScreen);
