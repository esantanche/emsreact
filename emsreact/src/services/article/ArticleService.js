// App configuration
import { APP_CONFIGURATION } from '../../appConfiguration';

// Helper function to calculate the number of the next page to fetch
// The first page is page zero
import { number_of_page_to_fetch } from '../../helpers/cache_computations';

/**
 * This service fetches articles from the backend.
 */
class ArticleService {

    constructor() {

        this.articles_cache = {};
        this.articles_count = {};

    }

    /**
     * Function getting articles.
     *
     * There are three implemented filters.
     *
     * The first one fetches a single article. It needs the node id the backend assigned to the article.
     *
     * The second fetches all articles for a given topic. Topics can be "How I work", "The tools I use", etc.
     *
     * The third fetches sticky articles only or all af them.
     *
     * It's possible to combine the topic and sticky filters. Of course the nid filter (single article) has to be used alone.
     *
     * @param {object} filters Filters to use when fetching articles
     * @param {function} callback_to_return_articles_and_more_flag Function to call to return the fetched articles
     */
    get_articles(filters, callback_to_return_articles_and_more_flag) {

        let filter_query_string = "";

        if (filters) {

            if (filters.nid)
                filter_query_string += "&nid=" + filters.nid;

            if (filters.topic)
                filter_query_string += "&field_ems_topic_target_id=" + filters.topic;

            if (filters.sticky)
                filter_query_string += "&sticky=1";

        }

        this.fetch_articles(filter_query_string, (articles_fetched_successfully) => {

            if (articles_fetched_successfully) {

                // I need to know if there are more articles to fetch so that I can return
                // this piece of information and the caller can show the LOAD MORE button

                const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);

                const count_of_articles = this.count_of_articles_for_query(filter_query_string);

                const there_are_more_articles_to_fetch = (count_of_articles > count_of_articles_in_cache);

                const result_to_return = {

                    more: there_are_more_articles_to_fetch,

                    articles: this.articles_cache[filter_query_string]

                };

                callback_to_return_articles_and_more_flag(result_to_return);

            }

        })

    };

    /**
     * Doing the actual fetch. This function will add to the cache an additional page of results or it will do nothing
     * if all articles have already been fetched.
     * If there is nothing in cache, it will fetch page zero.
     * This function returns just true on success and false on failure. The caller will find the articles in cache.
     *
     * @param {string} filter_query_string String with query parameters that will filter the result
     * @param {function} callback Function to call to return true if everything went well, false otherwise.
     */
    fetch_articles(filter_query_string, callback) {

        let page_number;

        // There is a separated cache for each query string
        const count_of_articles_in_cache = this.count_of_articles_cache_for_query(filter_query_string);
        const count_of_articles = this.count_of_articles_for_query(filter_query_string);

        // We need to fetch another page only if we didn't yet fetch all available articles for the
        // given query.
        // If the cache is empty certainly we need to fetch the first page
        if (count_of_articles > count_of_articles_in_cache || count_of_articles_in_cache === 0) {

            if (count_of_articles === -1)
                page_number = 0;
            else
                page_number = number_of_page_to_fetch(count_of_articles_in_cache, count_of_articles, APP_CONFIGURATION.fetchPageSize);

            fetch(APP_CONFIGURATION.backendUrl + "/rest/EMS/view/articlesplayground?_format=json&langcode=en" +
                        filter_query_string + "&page=" + page_number, {
                method: 'GET',
            })
                .then((response) => {
                    if (!response.ok) { throw response }
                    return response.json()
                })
                .then((response) => {

                    this.add_results_to_cache(filter_query_string, response.results);

                    this.update_count_of_articles_for_query(filter_query_string, parseInt(response.count));

                    callback(true);
                })
                .catch((error) => {

                    callback(false);
                });


        } else {

            // Everything is already in cache. The caller has just to read it.

            callback(true);

        }

    };

    /**
     * Returning the number of articles we already have in cache for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Number of articles in cache for the given query string
     */
    count_of_articles_cache_for_query(filter_query_string) {

        // Testing filter_query_string for empty
        // See https://repl.it/@EmanueleSantanc/Check-for-empty-parameter

        if (!filter_query_string) {

            console.error("count_of_articles_cache_for_query, filter_query_string not defined");

            return -1;
        }

        // If the cache for the given query string is not even defined,
        // we define it and set it as empty
        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        return this.articles_cache[filter_query_string].length;

    };

    /**
     * Returning the total number of articles that are available for the given
     * query string.
     * We don't get the number here. We got it previously and stored it in an array.
     * Now just returning the stored value.
     *
     * @param {string} filter_query_string Query string
     * @returns {number} Total number of articles that are available for the given
     * query string.
     */
    count_of_articles_for_query(filter_query_string) {

        if (!filter_query_string) {

            console.error("count_of_articles_for_query, filter_query_string not defined");

            return -1;
        }

        // If the total number of articles available for a given query has never been
        // determined, we define it and set it to -1
        if (typeof this.articles_count[filter_query_string] === "undefined") {

            this.articles_count[filter_query_string] = -1;

        }

        return this.articles_count[filter_query_string];

    };

    /**
     * We call this function to update the total number of articles available for a given query.
     * We typically do this after we make an actual query.
     *
     * @param {string} filter_query_string Query string
     * @param {number} count_of_articles Total number of articles that are available for the given
     * query string.
     */
    update_count_of_articles_for_query(filter_query_string, count_of_articles) {

        if (!filter_query_string) {

            console.error("update_count_of_articles_for_query, filter_query_string not defined");

            return;
        }

        this.articles_count[filter_query_string] = count_of_articles;

    };

    /**
     * Adding articles to the cache of articles for the given query string
     *
     * @param {string} filter_query_string Query string
     * @param {array} results Array of additional results to append to the cache
     */
    add_results_to_cache(filter_query_string, results) {

        if (typeof this.articles_cache[filter_query_string] === "undefined") {

            this.articles_cache[filter_query_string] = [];

        }

        // Append array to array
        // See https://repl.it/@EmanueleSantanc/Append-array-to-an-array

        this.articles_cache[filter_query_string].push(...results);

    }

}

export default ArticleService;