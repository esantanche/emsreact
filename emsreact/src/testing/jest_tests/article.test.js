// Here we test components in the folder article

import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import ArticleService from "../../services/article/ArticleService";
import AppContext from "../../AppContext";
import ArticleScreen from "../../article/ArticleScreen";

const AppArticleService = new ArticleService();

const Services = { TopicService: null,
    ArticleService: AppArticleService,
    PickMyBrainMessageService: null };

const json_response_rest_api_call = [
    {
        "nid": "58",
        "title": "Privacy Policy",
        "body": "This is the body of the privacy policy. It\' for testing",
        "langcode": "English",
        "field_ems_topic": "Small print",
        "sticky": "Off",
        "field_image": ""
    }
];

const article_screen_to_be_tested = ( <AppContext.Provider value={Services}>
                                         <ArticleScreen match={{ params: { articlenid: 58 } }}
                                         />
                                      </AppContext.Provider> );

describe('Testing ArticleScreen.js', function() {

    it('ArticleScreen renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(article_screen_to_be_tested , div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('ArticleScreen mocking fetch call', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call));

        const div = document.createElement('div');
        ReactDOM.render(article_screen_to_be_tested, div);
        ReactDOM.unmountComponentAtNode(div);

        // There has to be one fetch call only
        expect(fetch.mock.calls.length).toEqual(1);

        // The first argument in the fetch call has to be a string
        // containing rest/EMS/view/articles
        // The first argument is the rest api url
        expect(fetch.mock.calls[0][0]).toMatch(/rest\/EMS\/view\/articles/);

    });

    // This test is supposed to return a snapshot with no article because the
    // component ArticleScreen will be rendered before it updates with
    // the response of the fetch
    it('ArticleScreen renders no article before receiving fetch response', () => {

        fetch.resetMocks();

        fetch.mockResponseOnce(JSON.stringify(json_response_rest_api_call));

        const tree = renderer
            .create(article_screen_to_be_tested)
            .toJSON();
        expect(tree).toMatchSnapshot();

    });

});
