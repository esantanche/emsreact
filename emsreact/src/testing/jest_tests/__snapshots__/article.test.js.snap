// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ArticleScreen.js ArticleScreen renders no article before receiving fetch response 1`] = `
<div
  className="MuiGrid-container-357 MuiGrid-spacing-xs-16-379"
>
  <div
    className="MuiGrid-item-358 MuiGrid-grid-md-1-413 MuiGrid-grid-lg-2-428"
  />
  <div
    className="MuiGrid-item-358 MuiGrid-grid-xs-12-396 MuiGrid-grid-md-10-422 MuiGrid-grid-lg-8-434"
  >
    <div
      className="MuiPaper-root-454 MuiPaper-elevation1-457 MuiPaper-rounded-455 MuiCard-root-453"
    >
      <div
        className="MuiCardHeader-root-481"
      >
        <div
          className="MuiCardHeader-content-484"
        >
          <span
            className="MuiTypography-root-487 MuiTypography-headline-492 MuiCardHeader-title-485"
          />
        </div>
        <div
          className="MuiCardHeader-action-483"
        >
          <a
            className="MuiButtonBase-root-517 MuiIconButton-root-511"
            href="/"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
          >
            <span
              className="MuiIconButton-label-516"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-520"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g>
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </g>
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-527"
            />
          </a>
        </div>
      </div>
      <div
        className="MuiCardContent-root-534"
      >
        <div
          className="MuiTypography-root-487 MuiTypography-body1-496"
        >
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": undefined,
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-358 MuiGrid-grid-md-1-413 MuiGrid-grid-lg-2-428"
  />
</div>
`;
